File: packages/rate-limiter/src/algorithms/token-bucket.ts
import { DEFAULT_WINDOW_MS, MAX_BURST } from '@rateforge/types';

import type { RateLimitResult } from '@rateforge/types';

interface BucketState {
  tokens: number;
  lastRefill: number;
}

interface TokenBucketOptions {
  /**
   * Maximum number of tokens in the bucket (the steady-state limit).
   */
  capacity: number;
  /**
   * Window size in milliseconds over which capacity tokens are refilled.
   * Defaults to DEFAULT_WINDOW_MS.
   */
  windowMs?: number;
  /**
   * Optional function to override current time (useful for tests).
   */
  now?: () => number;
}

export class TokenBucket {
  private readonly capacity: number;
  private readonly windowMs: number;
  private readonly now: () => number;
  private readonly buckets = new Map<string, BucketState>();

  constructor(options: TokenBucketOptions) {
    this.capacity = options.capacity;
    this.windowMs = options.windowMs ?? DEFAULT_WINDOW_MS;
    this.now = options.now ?? (() => Date.now());
  }

  private getBucket(key: string): BucketState {
    const existing = this.buckets.get(key);
    const now = this.now();

    if (!existing) {
      const initial: BucketState = {
        tokens: this.capacity + MAX_BURST,
        lastRefill: now
      };
      this.buckets.set(key, initial);
      return initial;
    }

    const elapsed = now - existing.lastRefill;
    if (elapsed <= 0) {
      return existing;
    }

    const refillRatePerMs = this.capacity / this.windowMs;
    const refilled = existing.tokens + elapsed * refillRatePerMs;
    existing.tokens = Math.min(refilled, this.capacity + MAX_BURST);
    existing.lastRefill = now;

    return existing;
  }

  consume(key: string, cost: number): RateLimitResult {
    const now = this.now();
    const bucket = this.getBucket(key);

    const limit = this.capacity;
    const resetAt = bucket.lastRefill + this.windowMs;

    if (bucket.tokens >= cost) {
      bucket.tokens -= cost;

      return {
        allowed: true,
        limit,
        remaining: Math.max(0, Math.floor(bucket.tokens)),
        resetAt
      };
    }

    const retryAfterMs = Math.max(0, resetAt - now);

    return {
      allowed: false,
      limit,
      remaining: Math.max(0, Math.floor(bucket.tokens)),
      resetAt,
      retryAfterMs,
      reason: 'TOKEN_BUCKET_EXHAUSTED'
    };
  }
}

File: packages/rate-limiter/src/algorithms/token-bucket.lua
--[[
  Token Bucket — atomic check-and-decrement via Redis EVAL.

  KEYS[1]  = bucket key  (e.g. rateforge:rl:rule-id:user:ip:endpoint)
  ARGV[1]  = capacity       (integer, max tokens)
  ARGV[2]  = refill_rate    (tokens per second, float string)
  ARGV[3]  = cost           (integer, tokens to consume, usually 1)
  ARGV[4]  = now_ms         (current Unix time in milliseconds)
  ARGV[5]  = window_ms      (window size in ms — used to set TTL)

  Returns a three-element array: { allowed, remaining, reset_at_ms }
    allowed    = 1 if request is permitted, 0 if rejected
    remaining  = integer tokens left after this call
    reset_at_ms = epoch ms at which the bucket next resets (approx)
--]]

local key          = KEYS[1]
local capacity     = tonumber(ARGV[1])
local refill_rate  = tonumber(ARGV[2])   -- tokens/ms
local cost         = tonumber(ARGV[3])
local now_ms       = tonumber(ARGV[4])
local window_ms    = tonumber(ARGV[5])

-- Read existing state: { tokens, last_refill_ms }
local data = redis.call('HMGET', key, 'tokens', 'last_refill')

local tokens      = tonumber(data[1])
local last_refill = tonumber(data[2])

if tokens == nil then
  -- First request: initialise full bucket
  tokens      = capacity
  last_refill = now_ms
else
  -- Refill proportionally to elapsed time
  local elapsed = now_ms - last_refill
  if elapsed > 0 then
    tokens = tokens + elapsed * refill_rate
    if tokens > capacity then tokens = capacity end
    last_refill = now_ms
  end
end

local allowed   = 0
local remaining = math.floor(tokens)

if tokens >= cost then
  tokens    = tokens - cost
  remaining = math.floor(tokens)
  allowed   = 1
end

-- Persist new state with TTL = window_ms (rounded up to seconds)
local ttl_s = math.ceil(window_ms / 1000)
redis.call('HSET', key, 'tokens', tostring(tokens), 'last_refill', tostring(last_refill))
redis.call('EXPIRE', key, ttl_s)

local reset_at_ms = last_refill + window_ms

return { allowed, remaining, reset_at_ms }
